File: python/coloring/luby_jones.py
import random

class Coloring:

	def __init__(self, graph=None):
		self.graph = graph
		self.vertices = graph.vertices
		
	def color(self, verify=False, sort="id"):
		"""Color the graph and return how many unique colors were used."""
		
		colored = 0
		round = 0
	
		if sort == "id":
			sorted_vertices = sorted(self.vertices, key = lambda x : x)
		elif sort == "ascending":
			sorted_vertices = sorted(self.vertices, key = lambda x : len(self.vertices[x].adjacent_to))
		elif sort == "descending":
			sorted_vertices = sorted(self.vertices, key = lambda x : -len(self.vertices[x].adjacent_to))
		
		# Color all vertices
		while colored < len(sorted_vertices):
			
			# Test if vertex needs to be colored this round
			for vertex in sorted_vertices:
				if 'color' in self.graph.get_vertex_value(vertex):
					continue
			
				neighbors = self.graph.neighbors(vertex)
				local_max = True
				
				# Check if local max
				for neighbor in neighbors:
					data_neighbor = self.graph.get_vertex_value(neighbor)
					if 'color' in data_neighbor and data_neighbor['color'] != round:
						continue
					if neighbor > vertex:
						local_max = False
				
				# Color if local  max
				if local_max:
					self.graph.set_vertex_value(vertex, {u'color': round})
					colored += 1
								
			# Increment round
			round += 1
		
		# Verify
		correct = True
		if verify:
			for vertex in sorted_vertices:
				neighbors = self.graph.neighbors(vertex)
				for neighbor in neighbors:
					if self.graph.get_vertex_value(vertex)['color'] == self.graph.get_vertex_value(neighbor)['color']:
						return -1
		
		return round

File: cpp/GraphColoring/source/LubyJones.cpp
#include "..\include\LubyJones.h"

#include <iostream>

#define FALSE				0
#define TRUE				1
#define ALL					0

#define SKIP				0
#define UNKNOWN				1
#define LOCAL_MAX			2
#define NOT_LOCAL_MAX		3

// Constructor
LubyJones::LubyJones(Graph &graph, Graph::VertexVector &sortedVertices)
	: Colorer(graph, sortedVertices) { }

// Color the graph.
int LubyJones::color(bool verify, std::vector<uint16_t> params)
{
	// Grab parameters.
	uint16_t randomNumbers = params.at(0);
	uint16_t MISiteratios = params.at(1);

	// Return 0 for empty graph.
	if (sortedVertices.size() == 0)
		return 0;

	// Assign numbers...
	Graph::VertexVector vertexNumbers = Graph::VertexVector(sortedVertices.size());

	// sequentially, or
	if (params.at(0) == FALSE)
	{
		int number = sortedVertices.size();
		for (Graph::Vertex vertex : sortedVertices)
			vertexNumbers.at(vertex) = --number;
	}

	// randomly.
	else
	{
		// to do
	}
	
	// Color.
	uint64_t colored = 0;
	uint64_t toColor = sortedVertices.size();
	Graph::Color round = 0;

	// Main coloring loop.
	while (colored < toColor)
	{
		// Round number serves as color for this round (increment first to start at color 1).
		round++;
		
		// Maximal Independent Set algorithm parameters.
		Graph::VertexVector independentSet = Graph::VertexVector();
		uint64_t known = 0;
		uint64_t toKnow = 0;
		uint16_t iteration = 0;

		// Vertices to find MIS in this round.
		std::vector<uint16_t> vertices = std::vector<uint16_t>(sortedVertices.size());

		#pragma omp parallel for reduction(+:toKnow)
		for (int64_t vertex = 0; vertex < sortedVertices.size(); vertex++)
		{
			if (graph.getColor(vertex) == 0)
			{
				vertices.at(vertex) = UNKNOWN;
				toKnow++;
			}
		}
		
		// MIS main loop.
		while (known < toKnow && (iteration < MISiteratios || MISiteratios == ALL))
		{
			
			// Choose independent set in remaining vertices.
			uint64_t knownShared = 0;

			for (int64_t vertex = 0; vertex < sortedVertices.size(); vertex++)
			{
				
				// Skip if already colored.
				if (vertices.at(vertex) == SKIP || vertices.at(vertex) == NOT_LOCAL_MAX)
					continue;

				// Check if vertex is local max.
				Graph::Vertex vertexNumber = vertexNumbers.at(vertex);
				Graph::VertexVector neighbors = graph.getNeighbors(vertex);
				bool localMax = true;

				for (int64_t neighbor = 0; neighbor < neighbors.size(); neighbor++)
				{
					Graph::Color neighborColor = graph.getColor(neighbor);
					if (vertexNumbers.at(neighbor) > vertexNumber && vertices.at(neighbor) != SKIP)
					{
						localMax = false;
						break;
					}
				}

				// If local max, mark vertex as LOCAL_MAX and neighbors as NOT_LOCAL_MAX.
				if (localMax)
				{
					vertices.at(vertex) = LOCAL_MAX;
					knownShared++;

					for (Graph::Vertex neighbor : neighbors)
					{
						if (vertices.at(neighbor) != NOT_LOCAL_MAX && vertices.at(neighbor) != SKIP)
						{
							vertices.at(neighbor) = NOT_LOCAL_MAX;
							knownShared++;
						}
					}

					// Also add vertex to current round independent set.
					independentSet.push_back(vertex);
				}
			}

			// Mark current LOCAL_MAX and NOT_LOCAL_MAX as skippable for next MIS iteration.
			for (Graph::Vertex vertex = 0; vertex < vertices.size(); vertex++)
				if (vertices.at(vertex) == LOCAL_MAX || vertices.at(vertex) == NOT_LOCAL_MAX)
					vertices.at(vertex) = SKIP;

			known += knownShared;
			iteration++;
		}

		// Color independent set
		for (Graph::Vertex vertex : independentSet)
		{
			graph.setColor(vertex, round);
			colored++;
		}

	}
	
	// Verify and return.
	if (verify && !Colorer::verify())
		return -1;
		
	return round;
}